neural_compressor/torch/algorithms/weight_only/hqq/core.py

```diff
@@ -0,0 +1,273 @@
+                W_r = W_r.half()
+            if (meta["group_size"] is not None) and (meta["nbits"] == 3):
+                W_r = W_r[: meta["group_size"]] if (meta["axis"] == 0) else W_r[:, : meta["group_size"]]
+        else:
+            if hqq_global_option.use_half:
+                W_r = W_q.half()
+        # TODO: double check the correctness, the official impl is also error...
+        W_r = ((W_r - meta["zero"]) * meta["scale"]).reshape(meta["shape"])
+        return W_r
+
+
+class HQQLinear(torch.nn.Linear):
+    # The design follows the https://github.com/pytorch-labs/ao.
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = True,
+        q_weight: QTensor = None,
+        device=None,
+        dtype=None,
+    ) -> None:
+        super().__init__(in_features, out_features, bias, device, dtype)
+        self.q_weight = q_weight
+        self.quantized = q_weight is not None
+
+    @dump_elapsed_time("Quantize linear module into HQQ module.")
+    def quantize_weight(
+        self,
+        W: torch.Tensor,
+        quant_config: HQQModuleConfig = default_hqq_module_config,
+    ) -> Tuple[torch.Tensor, Dict[str, Any]]:
+        weight_quant_config, scale_quant_config, zero_quant_config = (
+            quant_config.weight,
+            quant_config.scale,
+            quant_config.zero,
+        )
+        need_quant_scale = scale_quant_config is not None
+        need_quant_zero = zero_quant_config is not None
+
+        self.in_features, self.out_features = W.t().shape
+
+        # Quantize weight
+        q_weight = HQQTensorHandle.quantize_to_q_tensor(float_tensor=W, tensor_quant_config=weight_quant_config)
+        self.q_weight = q_weight
+
+        # * The dequantization process only happens in the first forward pass.
+        # * It will change the `q_weight` but faster.
+        # * we should not save the state after doing the forward.
+        if need_quant_scale:  # Quantize scale
+            q_scale_tensor = HQQTensorHandle.quantize_to_q_tensor(
+                float_tensor=self.q_weight.scale, tensor_quant_config=scale_quant_config
+            )
+            self.q_weight.scale = q_scale_tensor
+        if need_quant_zero:  # Quantize zero
+            q_zero_tensor = HQQTensorHandle.quantize_to_q_tensor(
+                float_tensor=self.q_weight.zero,
+                tensor_quant_config=zero_quant_config,
+            )
+            self.q_weight.zero = q_zero_tensor
+        self.quantized = True
+
+    def dequantize_weight(self):
+        assert self.quantized, "model was not quantized"
+        # TODO: move below logic into `HQQTensorHandle`
+        if self.q_weight.is_scale_quantized():
+            scale_qdq = HQQTensorHandle.dequantize_q_tensor(self.q_weight.scale)
+            self.q_weight.scale = scale_qdq
+
+        if self.q_weight.is_zero_quantized():
+            zero_qdq = HQQTensorHandle.dequantize_q_tensor(self.q_weight.zero)
```

Comment:
have you synced with Xin for the export format

Reply:
Ok, I will sync with Xin regarding the export format. The `save`, `load`, and `export` functionalities will be in a separate PR.